### DISCUSSION TITLE
Migrate Cool Stuff from Orienteer to OrientDB

### DISCUSSION THREAD

**Discussion by PhantomYdn** [orientechnologies/orientdb#9560] [Ideas]
Hello,

We, [Orienteer](https://github.com/OrienteerBAP/Orienteer) (Business Application Platform), use OrientDB very deeply. Over time we have implemented several neat stuff which might be useful for OrientDB community as well. Propose to discuss in this thread things which can be migrated to OrientDB for wider use.

Orienteer DAO - OrientDB Objects and JPAs are too heavy and do not cover all OrientDB features. Orienteer DAO allows much easier creation and working with OrientDB data model. For example, it supports multiple parents per class, in place queries and etc. Check DAO classes here as an example: https://github.com/OrienteerBAP/Orienteer/tree/master/orienteer-core/src/test/java/org/orienteer/core/dao

OSchemaHelper - utility class which helps to create/update OrientDB data model in a declarative and idempotent way. 
[Example](https://github.com/OrienteerBAP/Orienteer/blob/be23f0c564331a448d383f22cd0033071c16ee76/orienteer-pages/src/main/java/org/orienteer/pages/module/PagesModule.java#L72-L80)

Orienteer Architect - UML like UI editor for OrientDB datamodel. 

![Orienteer Architect](https://raw.githubusercontent.com/OrienteerBAP/Orienteer/gh-pages/images/overview/Orienteer_Architect.png)

[Orienteer Camel](https://github.com/OrienteerBAP/camel-orientdb) - OrientDB adaptors for [Apache Camel ](https://camel.apache.org/)

[Orienteer BIRT](https://github.com/OrienteerBAP/orientdb-oda-birt) - OrientDB adaptors for [Eclipse BIRT](https://www.eclipse.org/birt/)

For sure, we have much more stuff. I will add more into this thread once "easily detachable" things can be identified.

**Comment by luigidellaquila** [comment]
Hi @PhantomYdn 

Thank you very much! 
I see a lot of cool things here that can definitely be useful to the community.
Some of them make a lot of sense as add-ons, some probably can also be considered as an integration to the standard distribution (eg. the SchemaHelper? Not sure if some of its functionalities relate to core features though, eg `assignVisualization()` ).

This is my personal opinion, but I'd really like to see further feedback from the Community here, to understand which of these components should have higher priority.
As you know, there will be no further SAP Enterprise OrientDB releases after v 3.2, so contributions from the Community become much more important for the evolution of OrientDB Community Edition.

Thanks!

Luigi

**Reply by luigidellaquila** [reply]
A few quick questions about Orienteer Architect: 
How is it implemented? Is it a normal Web App? Full client or also a server backend?
Does it infer the schema from the data or does it just read the schema itself?
Do you think it would be compatible with Studio architecture?

Thanks

Luigi

**Reply by PhantomYdn** [reply]
Orienteer Architect is primary Client App, but today it has some server-side. But it's not a rocket-science to re-implement that to fully OrientDB REST.

Regarding DAO: yes - you are right - currently there are some just Orienteer specific (like visualization), but it's even more easier detachable. And generalized processor for additional annotations might be helpful for other projects too.

**Comment by PhantomYdn** [comment]
Few more things to consider:

1) [Calculable Properties](https://github.com/OrienteerBAP/Orienteer/blob/master/orienteer-core/src/main/java/org/orienteer/core/hook/CalculablePropertiesHook.java). Hook which allows to define formula how value for a property should be calculated every time when document read.
2) [Reference Consistency](https://github.com/OrienteerBAP/Orienteer/blob/master/orienteer-core/src/main/java/org/orienteer/core/hook/ReferencesConsistencyHook.java) Hook which maintains links consistency.  For example: There is class `Folder` with property `files(LINKLIST)` which points to instances of class `File`. You can define property `folder(LINK)` on `File` which will be consistent with `files(LINKLIST)` and vice versa. For instance: you can create `File` which refers to some `Folder` and property `files` will be updates accordingly.
3) [Callback Hook](https://github.com/OrienteerBAP/Orienteer/blob/master/orienteer-core/src/main/java/org/orienteer/core/hook/CallbackHook.java) Hook which allows dynamically execute some code on document save and etc.
